Upgrade to 721Delegate v2

johnnyd-eth opened the pull request:
## What does this PR do and why?

Implements changes to 721Delegate as per https://juicebox.notion.site/JB721Delegate-v2-64d8d7435d8f4317bee29d07ce305c42. 

- Adds `preventOverspending` setting to create flow and launch tx 
<img width="632" alt="Screen Shot 2023-01-08 at 3 41 33 pm" src="https://user-images.githubusercontent.com/96150256/211182321-578edc56-9eb2-4135-a6cc-3bd322c5428a.png">

<img width="860" alt="Screen Shot 2023-01-08 at 3 41 38 pm" src="https://user-images.githubusercontent.com/96150256/211182323-b3c21def-f54b-4574-97ad-44b73e817a06.png">

- Add `flagsOf` contract read in `NftRewardsProvider` so we can get a project's `preventOverspending`
- If `preventOverspending` is enabled, we only select NFTs from the pay input when the amount matches the NFT's floor exactly. 

- Changes the `payMetadata` to have only allowOverspending(=true) and tierIdsToMint


## Screenshots or screen recordings

_If applicable, provide screenshots or screen recordings to demonstrate the changes._

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [ ] I have tested this PR in dark mode and light mode (if applicable).

tomquirk commented:
note just in case: we need to wait til the new version of contracts is deployed, and then update the version in package.json

tomquirk commented on src/components/v2v3/V2V3Project/V2V3FundingCycleSection/settingExplanations.tsx:
```diff
@@ -90,3 +90,11 @@ export const USE_DATASOURCE_FOR_REDEEM_EXPLAINATION = (
+export const PREVENT_OVERSPENDING_EXPLAINATION = (
+  <Trans>
+    When enabled, users can only receive NFTs by paying the exact price of the
+    NFT. This ensures that when users redeem their NFTs, they will receive the
+    amount they paid for them.
```
```suggestion
    NFT. This ensures contributors are eligible to receive their full payment amount if they chose to redeem their NFT.
```

johnnyd-eth replied:
Nice, done.

tomquirk commented on src/redux/slices/editingV2Project.ts:
```diff
@@ -3,13 +3,15 @@ import * as constants from '@ethersproject/constants'
 import { createSlice, PayloadAction } from '@reduxjs/toolkit'
 import { AllocationSplit } from 'components/Allocation'
 import { projectTokenSettingsToReduxFormat } from 'components/Create/utils/projectTokenSettingsToReduxFormat'
+import { DV_V1_1 } from 'constants/delegateVersions'
```
**suggestion** can we change the name of this to be more description `JB721DELEGATE_V1_1` maybe

johnnyd-eth replied:
Good call, done

tomquirk commented on src/redux/slices/editingV2Project.ts:
```diff
@@ -168,7 +178,8 @@ const defaultProjectState: ProjectState = {
     postPayModal: undefined,
-    contractVersion: undefined,
+    flags: DEFAULT_NFT_FLAGS,
+    contractVersion: DV_V1_1,
```
This isn't true yet.

johnnyd-eth replied:
True true, put it back to DV_V1_1.

tomquirk reviewed (commented):
@johnnyd-eth how relevant is this flag concept to the V1 delegate? Im worried about the bug risk here of trying to feature flag and interop the new flags with the old. What's your feeling?

johnnyd-eth commented:
Thanks @tomquirk I've addressed those comments of yours.

Additionally, in line with you moving the `useDataSourceForRedeem` stuff from the 'Rules' section to the 'NFT' section, I've done the same with the `preventOverspending` field:

<img width="651" alt="Screen Shot 2023-02-02 at 4 41 51 pm" src="https://user-images.githubusercontent.com/96150256/216252148-52cf2e67-f79b-4f98-bf63-1efd4dd90979.png">

<img width="651" alt="Screen Shot 2023-02-02 at 4 41 51 pm" src="https://user-images.githubusercontent.com/96150256/216252148-52cf2e67-f79b-4f98-bf63-1efd4dd90979.png">

(Note: it's still hidden until we flip `PREVENT_OVERSPENDING_FIELD_VISIBLE` when we bump to delegate `4.0.1`)

tomquirk commented on src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:
```diff
@@ -18,9 +18,14 @@ import { RewardsList } from '../../RewardsList'
+} from 'components/v2v3/V2V3Project/V2V3FundingCycleSection/settingExplanations'
 import { JuiceSwitch } from 'components/JuiceSwitch'
 
+export const PREVENT_OVERSPENDING_FIELD_VISIBLE = true
```
@johnnyd-eth does this need feature flag?
